src/libdrakvuf/vmi.c

```diff
@@ -1365,6 +1423,39 @@ void drakvuf_loop(drakvuf_t drakvuf, bool (*is_interrupted)(drakvuf_t, void*), v
     PRINT_DEBUG("DRAKVUF polling loop finished\n");
 }
 
+void drakvuf_toggle_context_based_interception(drakvuf_t drakvuf)
+{
+    bool toggle = !drakvuf->enable_cr3_based_interception;
+    status_t status;
+    vmi_pause_vm(drakvuf->vmi);
+
+    if(toggle)
+    {
+        status = vmi_slat_switch(drakvuf->vmi, drakvuf->altp2m_idrx);
+        if (VMI_FAILURE == status)
+            PRINT_DEBUG("Enabling context based interception failed. \n");
+
+        if ( !drakvuf->cr3 && !control_cr3_trap(drakvuf, 1) )
+            PRINT_DEBUG("Failed to enable CR3 trap for context based interception. \n");
+
+
+        drakvuf->enable_cr3_based_interception = true;
+    }
+    else
+    {
+        status = vmi_slat_switch(drakvuf->vmi, drakvuf->altp2m_idx);
+        if (VMI_FAILURE == status)
+            PRINT_DEBUG("Disabling context based interception failed. \n");
+
+        if ( !drakvuf->cr3 )
+            control_cr3_trap(drakvuf, 0);
```

Comment:
You shouldn't use `control_cr3_trap` directly. You should use and internal `drakvuf_trap` for that, so that this won't interfere with plugins that may also subscribe to cr3 traps.

Reply:
I'm not sure if this would interfere with other subscribers as it does check if there is registered traps in `drakvuf->cr3` in case there is it wouldn't disable the cr3_event, and will only be disabled if the list is empty and there are no subscribers.

On the other hand , using `drakvuf_add_trap` would require a separate callback for the cr3_interception check, as well as in the process of handling a cr3 event, it would invoke the callback to its relative position in the GSList, which could be in the middle, in that case it may affect the response flags (not necessarily), tho it is your decision to make.

Reply:
I see, OK